velox/functions/prestosql/StringFunctions.cpp

```diff
@@ -275,6 +275,174 @@ class ConcatFunction : public exec::VectorFunction {
   std::vector<StringView> constantStringViews_;
 };
 
+/**
+ * concat_ws(separator, string1, ..., stringN) → varchar
+ * Returns the concatenation of string1, string2, ..., stringN connected by
+ * separator.
+ * */
+class ConcatWsFunction : public exec::VectorFunction {
+ public:
+  ConcatWsFunction(
+      const std::string& name,
+      const std::vector<exec::VectorFunctionArg>& inputArgs) {
+    auto numArgs = inputArgs.size();
+    VELOX_USER_CHECK(
+        numArgs >= 2,
+        "{} requires 2 arguments at least, but got {}",
+        name,
+        numArgs);
+
+    BaseVector* constantPattern = inputArgs[0].constantValue.get();
+    VELOX_USER_CHECK(
+        constantPattern != nullptr && !constantPattern->isNullAt(0),
+        "{} requires first argument to be a constant string",
+        name);
+
+    delim_ =
+        constantPattern->as<ConstantVector<StringView>>()->valueAt(0).str();
+
+    // Save constant values to constantStrings_.
+    // Identify and combine consecutive constant inputs.
+    argMapping_.reserve(numArgs - 1);
+    constantStrings_.reserve(numArgs - 1);
+
+    for (auto i = 1; i < numArgs; ++i) {
+      argMapping_.push_back(i);
+
+      const auto& arg = inputArgs[i];
+      if (arg.constantValue) {
+        std::string value = arg.constantValue->as<ConstantVector<StringView>>()
+                                ->valueAt(0)
+                                .str();
+
+        column_index_t j = i + 1;
+        for (; j < inputArgs.size(); ++j) {
+          if (!inputArgs[j].constantValue) {
+            break;
+          }
+
+          value += delim_ +
+              inputArgs[j]
+                  .constantValue->as<ConstantVector<StringView>>()
+                  ->valueAt(0)
+                  .str();
+        }
+
+        constantStrings_.push_back(std::string(value.data(), value.size()));
+
+        i = j - 1;
+      } else {
+        constantStrings_.push_back(std::string());
+      }
+    }
+
+    // Create StringViews for constant strings.
+    constantStringViews_.reserve(numArgs - 1);
+    for (const auto& constantString : constantStrings_) {
+      constantStringViews_.push_back(
+          StringView(constantString.data(), constantString.size()));
+    }
+  }
+
+  bool propagateStringEncodingFromAllInputs() const override {
+    return true;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, VARCHAR(), result);
+    auto flatResult = result->asFlatVector<StringView>();
+
+    auto numArgs = argMapping_.size();
+
+    std::vector<exec::LocalDecodedVector> decodedArgs;
+    decodedArgs.reserve(numArgs);
+
+    for (auto i = 0; i < numArgs; ++i) {
```

Comment:
Separate the constant mapping and flat vector as this one, https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/DecimalVectorFunctions.cpp#L50